sklearn/metrics/_ranking.py

```diff
@@ -1500,21 +1508,42 @@ def _dcg_sample_scores(y_true, y_score, k=None, log_base=2, ignore_ties=False):
         Cumulative Gain (the DCG obtained for a perfect ranking), in order to
         have a score between 0 and 1.
     """
-    discount = 1 / (np.log(np.arange(y_true.shape[1]) + 2) / np.log(log_base))
+    xp, _, device_ = get_namespace_and_device(y_true, y_score)
+    float_dtype = _find_matching_floating_dtype(y_true, y_score, xp=xp)
+    discount = 1 / (
+        xp.log(xp.astype(xp.arange(y_true.shape[1], device=device_), float_dtype) + 2)
+        / xp.log(xp.asarray(log_base, dtype=float_dtype, device=device_))
+    )
     if k is not None:
         discount[k:] = 0
     if ignore_ties:
-        ranking = np.argsort(y_score)[:, ::-1]
-        ranked = y_true[np.arange(ranking.shape[0])[:, np.newaxis], ranking]
-        cumulative_gains = discount.dot(ranked.T)
+        if _is_numpy_namespace(xp):
+            ranking = np.argsort(y_score)[:, ::-1]
+            ranked = y_true[np.arange(ranking.shape[0])[:, np.newaxis], ranking]
+            cumulative_gains = discount.dot(ranked.T)
+        else:
+            ranking = xp.flip(xp.argsort(y_score), axis=1)
+            # TODO this doesn't work yet for array_api_strict
+            ranking_index = xp.arange(ranking.shape[0])[:, xp.newaxis]
+            ranked = y_true[ranking_index, ranking]
+            cumulative_gains = xp.tensordot(discount, ranked.T, axes=1)
     else:
-        discount_cumsum = np.cumsum(discount)
-        cumulative_gains = [
-            _tie_averaged_dcg(y_t, y_s, discount_cumsum)
-            for y_t, y_s in zip(y_true, y_score)
-        ]
-        cumulative_gains = np.asarray(cumulative_gains)
-    return cumulative_gains
+        if _is_numpy_namespace(xp):
+            discount_cumsum = np.cumsum(discount)
+            cumulative_gains = [
+                _tie_averaged_dcg(y_t, y_s, discount_cumsum)
+                for y_t, y_s in zip(y_true, y_score)
```

Comment:
Won't something like this work generically?
```suggestion
                for y_t, y_s in xp.stack((y_true, y_score), axis=1)
```

Reply:
this seems to make `array_api_strict` complain if `y_true` and `y_score` don't have the same type (like in this test case): `TypeError: array_api_strict.int64 and array_api_strict.float64 cannot be type promoted together`

(`array_api_strict` complained a lot while working on this 😅)